File: 프로그래머스/lv3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
#include <iostream>

using namespace std;


long long max_T(vector<int> times_){
    int max_t = times_[0];
    for(int elem:times_){
        if(elem > max_t)
            max_t = elem;
    }
    return max_t;
}

long long get_person(vector<int> tim, long long mid){
    long long person=0;
    for(int i=0;i<tim.size();i++){
        person += mid/tim[i];
    }
    return person;
}
long long solution(int n, vector<int> times) {
    long long answer = 0;
    long long MAXT = max_T(times)*n;
    long long MINT = 0;
    while(MINT<=MAXT){
        long long mid = (MINT + MAXT)/2;
        long long person=get_person(times, mid);
        
        if(person >= n)
            MAXT = mid-1;
        else if(person < n)
            MINT = mid+1;
    }
    
    return MAXT+1;
}

File: 프로그래머스/lv3/43238. 입국심사/입국심사.java


import java.io.*;
import java.util.*;

public class Solution {
    public long solution(int n, int[] times) {
        long answer = 1;
        int length = times.length;
        Arrays.sort(times);

        while(n>0){
            for(int i=0;i<length;i++){
                if(answer % times[i] == 0)
                    n--;
            }
            answer++;
        }

        return answer-1;
    }
}